### PR TITLE
feat: changing cycles to discount in subscriptions and plans

### PR DESCRIPTION
**Pull request by mateus-picoloto**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| **Issue**         | https://mundipagg.atlassian.net/browse/PAOPN-575
| **What?**         | Changing cycles to discount in subscriptions and plans.
| **Why?**          | To be able to apply discount in all charges from subscriptions and plans.
| **How?**          | Changing cycles to discount in subscriptions and plans.

### CODE REVIEW

**Review by RafaMelazzo — CHANGES_REQUESTED** [review]

File: etc/adminhtml/system/recurrence.xml [hunk 1]
```diff
@@ -64,20 +64,22 @@
                 <field id="active">1</field>
             </depends>
         </field>
-        <field id="add_shipping_in_items" showInDefault="1" showInStore="1" showInWebsite="1" sortOrder="20" translate="label" type="select">
+        <field id="add_shipping_in_items" showInDefault="1" showInStore="1" showInWebsite="1" sortOrder="20" translate="label tooltip" type="select">
             <label>Add shipping values in items</label>
             <comment />
             <source_model>Magento\Config\Model\Config\Source\Yesno</source_model>
             <config_path>pagarme_pagarme/recurrence/add_shipping_in_items</config_path>
+            <tooltip>For plans, when enabling this options, the shipping it will apply to all charges even if the products have different cycles.</tooltip>
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply etc/adminhtml/system/recurrence.xml]: Ajustado

File: etc/adminhtml/system/recurrence.xml [hunk 2]
```diff
@@ -64,20 +64,22 @@
...
                 <field id="active">1</field>
             </depends>
         </field>
-        <field id="add_tax_in_items" showInDefault="1" showInStore="1" showInWebsite="1" sortOrder="20" translate="label" type="select">
+        <field id="add_tax_in_items" showInDefault="1" showInStore="1" showInWebsite="1" sortOrder="20" translate="label tooltip" type="select">
             <label>Add tax values in items</label>
             <comment />
             <source_model>Magento\Config\Model\Config\Source\Yesno</source_model>
             <config_path>pagarme_pagarme/recurrence/add_tax_in_items</config_path>
+            <tooltip>For plans, when enabling this options, the taxes it will apply to all charges even if the products have different cycles.</tooltip>
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply etc/adminhtml/system/recurrence.xml]: Ajustado

File: etc/db_schema.xml [hunk 3]
```diff
@@ -45,6 +45,28 @@
         <column name="cycles" nullable="true" xsi:type="int" padding="11" comment="Cycles" identity="false"/>
         <column name="created_at" nullable="false" xsi:type="timestamp" comment="created_at" default="CURRENT_TIMESTAMP"/>
         <column name="updated_at" nullable="false" xsi:type="timestamp" comment="updated_at" default="CURRENT_TIMESTAMP" on_update="true"/>
+        <column name="apply_products_cycle_to_discount" nullable="true" xsi:type="smallint" comment="Apply products cycle to discount" padding="6"/>
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply etc/db_schema.xml]: Ajustado

File: etc/db_schema.xml [hunk 4]
```diff
@@ -45,6 +45,28 @@
...
+        <constraint xsi:type="primary" referenceId="PRIMARY">
+            <column name="id"/>
+        </constraint>
+    </table>
+    <table name="pagarme_module_core_recurrence_products_plan" resource="default" engine="innodb" comment="Pagarme Recurrence Products Plan Table">
+        <column xsi:type="int" name="id" padding="10" unsigned="true" nullable="false" identity="true" comment="ID"/>
+        <column name="interval_type" nullable="false" xsi:type="varchar" comment="Day, week, month ou year" length="15"/>
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply etc/db_schema.xml]: Ajustado

File: etc/db_schema.xml [hunk 5]
```diff
@@ -45,6 +45,28 @@
...
+        <column name="interval_count" nullable="false" xsi:type="smallint" comment="Interval Count" padding="6" identity="false"/>
+        <column name="name" nullable="true" xsi:type="varchar" comment="Product name" length="255"/>
+        <column name="description" nullable="true" xsi:type="text" comment="Product description"/>
+        <column name="plan_id" nullable="true" xsi:type="varchar" comment="Api's id" length="21"/>
+        <column name="product_id" nullable="true" xsi:type="int" comment="Product in Magento's table" padding="11"/>
+        <column name="credit_card" nullable="false" xsi:type="char" comment="Accepts credit card" length="1"/>
+        <column name="installments" nullable="false" xsi:type="char" comment="Accepts installments" length="1"/>
+        <column name="boleto" nullable="false" xsi:type="char" comment="Accepts boleto" length="1"/>
+        <column name="billing_type" nullable="false" xsi:type="varchar" comment="Prepaid, postpaid ou exact_day" length="11"/>
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply etc/db_schema.xml]: Ajustado

File: i18n/pt_BR.csv [hunk 6]
```diff
@@ -499,3 +499,22 @@
...
 "Add tax values in items","Realizar a cobrança das taxas"
 "Taxs","Taxas"
+"Error in trying to create a plan type product","Erro ao tentar criar um produto do tipo plano"
+"Please add subproducts before plan saving","Por favor adicione subprodutos antes de salvar o plano"
+"Plan product not found","Produto do plano não encontrado"
+"List of plan products not found","Lista de produtos do plano não encontrado"
+"Plan Product saved","Produto do Plano salvo"
+"Subscription Product not found","Produto de assinatura não encontrado"
+"Subscription Products not found","Produtos de assinatura não encontrados"
+"Error on save subscription product","Erro ao salvar produto de assinatura"
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply i18n/pt_BR.csv]: Ajustado

File: i18n/en_US.csv [hunk 7]
```diff
@@ -317,3 +317,22 @@
...
+"Subscription product saved","Subscription product saved"
+"No product founded with name: %1","No product founded with name: %1"
+"Bundle product not selected","Bundle product not selected"
+"Select at last one payment method","Select at last one payment method"
+"Fill at last one cycle option","Fill at last one cycle option"
+"It was not possible to find the subproducts for this bundle. Check your configuration and try again","It was not possible to find the subproducts for this bundle. Check your configuration and try again"
+"Cart Discount","Cart Discount"
+"Apply cart discount to all subscription charges","Apply cart discount to all subscription charges"
+"For plans, when enabling this options, the discount will apply to all charges even if the products have different cycles.","For plans, when enabling this options, the discount will apply to all charges even if the products have different cycles."
+"For plans, when enabling this options, the shipping it will apply to all charges even if the products have different cycles.","For plans, when enabling this options, the shipping it will apply to all charges even if the products have different cycles."
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply i18n/en_US.csv]: Ajustado

File: i18n/en_US.csv [hunk 8]
```diff
@@ -317,3 +317,22 @@
...
+"For plans, when enabling this options, the taxes it will apply to all charges even if the products have different cycles.","For plans, when enabling this options, the taxes it will apply to all charges even if the products have different cycles."
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply i18n/en_US.csv]: Ajustado

File: i18n/pt_BR.csv [hunk 9]
```diff
@@ -499,3 +499,22 @@
...
+"Bundle product not selected","Produto bundle não selecionado"
+"Select at last one payment method","Selecione pelo menos um método de pagamento"
+"Fill at last one cycle option","Preencha pelo menos uma opção de pagamento"
+"It was not possible to find the subproducts for this bundle. Check your configuration and try again","Não foi possível encontrar os subprodutos deste bundle. Verifique sua configuração e tente novamente"
+"Cart Discount","Desconto de Carrinho"
+"Apply cart discount to all subscription charges","Aplicar desconto de carrinho para todas as cobranças da assinatura"
+"For plans, when enabling this options, the discount will apply to all charges even if the products have different cycles.","Para planos, ao habilitar essa opção, o desconto será aplicado a todas as cobranças da assinatura mesmo que os produtos tenham ciclos diferentes."
+"For plans, when enabling this options, the shipping it will apply to all charges even if the products have different cycles.","Para planos, ao habilitar essa opção, o frete será aplicado a todas as cobranças da assinatura mesmo que os produtos tenham ciclos diferentes."
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply i18n/pt_BR.csv]: Ajustado

File: i18n/pt_BR.csv [hunk 10]
```diff
@@ -499,3 +499,22 @@
...
+"For plans, when enabling this options, the taxes it will apply to all charges even if the products have different cycles.","Para planos, ao habilitar essa opção, as taxas serão aplicadas a todas as cobranças da assinatura mesmo que os produtos tenham ciclos diferentes."
```
[the remark this answers is not in the document]
Comment (mateus-picoloto) [reply i18n/pt_BR.csv]: Ajustado

**Review by RafaMelazzo — APPROVED** [review]

**Review by fabiano-mallmann — APPROVED** [review]

**Pull request merged**